README.md
# Wechat Swipeout

Inspired by [miniprogram-slide-view](https://github.com/wechat-miniprogram/slide-view), [react-native-swipeout](https://github.com/dancormier/react-native-swipeout).

## DEMO

### 手动关闭 or 自动关闭

- 第一个按钮手动地调用了 swipeout 的 `close` 方法
- 第二个 swipeout 设置了 `auto-close="true"`
- 最后一个 swipeout 什么也没做

![手动关闭 or 自动关闭](gif/manually-vs-automatically.gif)

### 吸附效果

1. 第一个 swipeout：左不吸附，右吸附
2. 第二个 swipeout：左不吸附，右不吸附
3. 第三个 swipeout：左吸附，右不吸附
4. 第四个 swipeout：左吸附，右吸附

![四个 swipeout 的左侧](gif/four-left-effects.gif) ![四个 swipeout 的右侧](gif/four-right-effects.gif)

### 摩擦系数

摩擦系数决定滑动速度：

![摩擦系数](gif/openleft-openright-close.gif)

### 禁用效果

设置了 `disabled="true"` 的 swipeout，无论开发者调用暴露的“打开”／“关闭”方法，还是用户触摸拖动滑块，都无效：

![禁用效果](gif/disabled.gif)

### 探索更多

以上仅演示部分效果，更多效果可以本地运行 demo 来体验：

1. `git clone git@github.com:baishusama/wechat-swipeout.git && cd wechat-swipeout && npm i && npm run dev`
2. 使用微信开发者工具打开 `./miniprogram_dev/` 这个 demo 小程序
3. enjoy it!

## 使用方法

> 关于小程序自定义组件的使用，详见[官方项目说明](https://github.com/wechat-miniprogram/miniprogram-custom-component)。

1. 安装 wechat-swipeout ：

```
npm i -S wechat-swipeout
```

2. 微信开发者工具 -> 工具 -> 构建 npm => 这步会生成 `miniprogram_npm/wechat-swipeout/index.(js|json|wxml|wxss)`
3. 在需要使用本组件的页面/组件的 `*.json` 文件中，设置组件 alias 和路径：

```json
{
  "usingComponents": {
    "swipeout": "wechat-swipeout"
  }
}
```

4. `*.wxml` 文件中，可以使用 `swipeout`：

```html
<swipeout class="swipeout" height="120" right-buttons="{{rightButtonList}}" auto-close="true" background-color="black" bind:press="onSwipeoutPress">
  <view class="main-content">Swipeout: 右侧按钮 & 自动关闭</view>
</swipeout>
```

更多示例，请参考 [github repo](https://github.com/baishusama/wechat-swipeout) 的 [`tools/demo/pages/index/index.wxml`](https://github.com/baishusama/wechat-swipeout/blob/master/tools/demo/pages/index/index.wxml) 文件内容。

具体接口参见 [API](#api) 小节。

## API

### 组件整体

#### 属性

| 属性 | 类型 | 可选 | 默认值 | 描述 |
|:-----|:----:|:---:|:-----:|:-----|
| `width` | number | 可选 | `750` | 主区域宽度 | 
| `height` | number | 可选 | `0` | 高度 |
| `background-color` | string | 可选 | `''` | 背景色 |
| `auto-close` | boolean | 可选 | `false` | 是否点击按钮后自动关闭 |
| `disabled` | boolean | 可选 | `false` | 禁用状态 |
| `damping` | number | 可选 | `30` | 阻尼系数，控制动画和过界回弹动画；越大越快 |
| `friction` | number | 可选 | `10` | 摩擦系数，控制惯性滑动；越大越快 |

#### 事件

| 事件 | 类型 | 可选 | 默认值 | 描述 |
|:-----|:----:|:---:|:-----:|:-----|
| `press` | method name | 可选 | - | 点击按钮时触发，会回传按钮的 hash 字段 | 

#### 暴露的方法

- `close` 关闭 swipeout
- `openLeft` 打开 swipeout 的左侧按钮
- `openRight` 打开 swipeout 的右侧按钮

### 两侧按钮

| 属性 | 类型 | 可选 | 默认值 | 描述 |
|:-----|:----:|:---:|:-----:|:-----|
| `(left\|right)ButtonWidth` | number | 可选 | `0` | 左／右侧每个按钮的宽度 | 
| `(left\|right)Width` | number | 可选 | `0` | 左／右侧所有按钮的宽度之和 | 
| `(left\|right)Buttons` | SwipeoutButton[] | 可选 | `null` | 左／右侧按钮列表 | 
| `stickTo(Left\|Right)Edge` | boolean | 可选 | `false` | 左／右侧按钮是否吸附边缘 | 

#### 关于 `SwipeoutButton` 类型

| 属性 | 类型 | 可选 | 默认值 | 描述 |
|:-----|:----:|:---:|:-----:|:-----|
| `hash` | string | 必须 | - | 按钮的唯一标识；点击按钮（`press`）时会回传 hash，可以此区分按钮并执行对应事件 | 
| `className` | string | 可选 | - | 按钮的类名，可用于微调样式 |
| `width` | number | 可选 | - | 按钮的宽度，单位 rpx |
| `iconPath` | string | 可选 | - | 按钮的 icon 的路径；注意，组件构建后位于 `miniprogram_npm/` 目录下，设置 `iconPath` 时需要考虑进去 |
| `text` | string | 可选 | - | 按钮的文字 |
| `color` | string | 可选 | - | 按钮的字体颜色 |
| `backgroundColor` | string | 必须 | - | 按钮的背景色 |
| `underlayColor` | string | 必须 | - | 按钮被点击时的背景色 |

## License

This project is licensed under the MIT License


CHANGELOG.md
# 1.1.1 (2018-12-13)

### Bug Fixes

* `npm publish` with `npm run build` at first

## 1.1.0 (2018-12-12)

### Bug Fixes

* fix: use buttons and enable customizing its `underlayColor` to avoid the color toggling bug brought by simulating buttons through view
* fix: overflow of content inside the area
* fix: apply prefix `wechat-swipeout__` to all classNames

## Before 1.1.0

Mysterious old versions:

- 1.0.4
- 1.0.3
- 1.0.2
- 1.0.1
- 1.0.0

.eslintrc.js
module.exports = {
  'extends': [
    'airbnb-base',
    'plugin:promise/recommended'
  ],
  'parserOptions': {
    'ecmaVersion': 9,
    'ecmaFeatures': {
      'jsx': false
    },
    'sourceType': 'module'
  },
  'env': {
    'es6': true,
    'node': true,
    'jest': true
  },
  'plugins': [
    'import',
    'node',
    'promise'
  ],
  'rules': {
    'arrow-parens': 'off',
    'comma-dangle': [
      'error',
      'only-multiline'
    ],
    'complexity': ['error', 15],
    'func-names': 'off',
    'global-require': 'off',
    'handle-callback-err': [
      'error',
      '^(err|error)$'
    ],
    'import/no-unresolved': [
      'error',
      {
        'caseSensitive': true,
        'commonjs': true,
        'ignore': ['^[^.]']
      }
    ],
    'import/prefer-default-export': 'off',
    'linebreak-style': 'off',
    'no-catch-shadow': 'error',
    'no-continue': 'off',
    'no-div-regex': 'warn',
    'no-else-return': 'off',
    'no-param-reassign': 'off',
    'no-plusplus': 'off',
    'no-shadow': 'off',
    'no-multi-assign': 'off',
    'no-underscore-dangle': 'off',
    'node/no-deprecated-api': 'error',
    'node/process-exit-as-throw': 'error',
    'object-curly-spacing': [
      'error',
      'always'
    ],
    'object-curly-newline': [
      'error',
      {
        "ObjectPattern": { "multiline": true },
        "ImportDeclaration": "never",
        "ExportDeclaration": { "multiline": true, "minProperties": 3 }
      }
    ],
    'operator-linebreak': [
      'error',
      'after',
      {
        'overrides': {
          ':': 'before',
          '?': 'before'
        }
      }
    ],
    'prefer-arrow-callback': 'off',
    'prefer-destructuring': 'off',
    'prefer-template': 'off',
    'quote-props': [
      1,
      'as-needed',
      {
        'unnecessary': true
      }
    ],
    'semi': [
      'error',
      'never'
    ]
  },
  'globals': {
    'window': true,
    'document': true,
    'App': true,
    'Page': true,
    'Component': true,
    'Behavior': true,
    'wx': true,
    'getCurrentPages': true,
  }
}


tools/config.js
const path = require('path')

const webpack = require('webpack')
const nodeExternals = require('webpack-node-externals')

const isDev = process.argv.indexOf('--develop') >= 0
const isWatch = process.argv.indexOf('--watch') >= 0
const demoSrc = path.resolve(__dirname, './demo')
const demoDist = path.resolve(__dirname, '../miniprogram_dev')
const src = path.resolve(__dirname, '../src')
const dev = path.join(demoDist, 'components')
const dist = path.resolve(__dirname, '../miniprogram_dist')

module.exports = {
  entry: ['index'],

  isDev,
  isWatch,
  srcPath: src,
  distPath: isDev ? dev : dist,

  demoSrc,
  demoDist,

  wxss: {
    less: true, // compile wxss with less
    sourcemap: false, // source map for less
  },

  webpack: {
    mode: 'production',
    output: {
      filename: '[name].js',
      libraryTarget: 'commonjs2',
    },
    target: 'node',
    externals: [nodeExternals()], // ignore node_modules
    module: {
      rules: [
        {
          test: /\.js$/i,
          use: ['babel-loader', 'eslint-loader'],
          exclude: /node_modules/,
        },
      ],
    },
    resolve: {
      modules: [src, 'node_modules'],
      extensions: ['.js', '.json'],
    },
    plugins: [
      new webpack.DefinePlugin({}),
      new webpack.optimize.LimitChunkCountPlugin({ maxChunks: 1 }),
    ],
    optimization: {
      minimize: false,
    },
    // devtool: 'nosources-source-map', // source map for js
    performance: {
      hints: 'warning',
      assetFilter: assetFilename => assetFilename.endsWith('.js'),
    },
  },

  copy: ['./wxml', './wxss', './wxs', './images'], // will copy to dist
}


src/index.js
const prefix = 'wechat-swipeout__'
const MIN_BTN_W = 100 // (rpx)
const WINDOW_W = wx.getSystemInfoSync().windowWidth // (px)

Component({
  // 选项
  options: {
    multipleSlots: true, // 使用多个 slot 时必须
    addGlobalClass: true
  },
  // 接口 (rpx)
  properties: {
    /* 主要内容 */
    // 组件显示区域的宽度，默认全屏宽度
    width: {
      type: Number,
      value: 750,
    },
    // 组件显示区域的高度
    height: {
      type: Number,
      value: 0,
    },
    /* 按钮相关 */
    // （左侧）按钮总宽度
    leftWidth: {
      type: Number,
      value: 0,
    },
    // （左侧）作为按钮的默认宽度
    leftButtonWidth: {
      type: Number,
      value: 0,
    },
    // （左侧）按钮列表
    leftButtons: {
      type: Array,
      value: null,
    },
    // （右侧）按钮总宽度
    rightWidth: {
      type: Number,
      value: 0,
    },
    // （右侧）作为按钮的默认宽度
    rightButtonWidth: {
      type: Number,
      value: 0,
    },
    // （右侧）按钮列表
    rightButtons: {
      type: Array,
      value: null,
    },
    // 贴边选项
    stickToLeftEdge: {
      type: Boolean,
      value: false
    },
    stickToRightEdge: {
      type: Boolean,
      value: false
    },
    /* 样式/状态 */
    backgroundColor: {
      type: String,
      value: ''
    },
    disabled: {
      type: Boolean,
      value: false
    },
    /* 行为表现 */
    autoClose: {
      type: Boolean,
      value: false
    },
    damping: {
      type: Number,
      value: 30
    },
    friction: {
      type: Number,
      value: 10
    },
  },
  data: {
    totalWidth: WINDOW_W, // (rpx) （包括未显示的按钮的）总宽度，默认全屏宽度
    leftVisibleWidth: 0, // (px) 左侧可视宽度
    rightVisibleWidth: 0, // (px) 右侧可视宽度
    isLeftSticked: false, // (px) 左侧吸边状态
    isRightSticked: false, // (px) 右侧吸边状态
    offsetX: 0, // (px) 默认偏移位置
    useAnimation: false,
    canBeOut: false, // 是否允许出界
  },
  lifetimes: {
    attached() {
      this.initializeWidth()
      this.calculateTotal()
      this.prepareThreshold()

      // 初始化 offsetX
      if (this.data.leftWidth) {
        this.setData({
          offsetX: -this.data.leftWidth / 750 * WINDOW_W
        })
      }
    },
    ready() {
      this.setupAnimation()
    }
  },
  methods: {
    // 初始化宽度 FIXME: bug ??? 需要一个 setter 监听？再自动调用？
    initializeWidth() {
      // type Direction = 'left' | 'right'
      const _initialize = direction => {
        const widthKey = `${direction}Width`
        const buttonWidthKey = `${direction}ButtonWidth`
        const buttonsKey = `${direction}Buttons`
        const getInitData = defaultWidth => {
          let sideTotalWidth = 0

          const buttons = this.data[buttonsKey].map(button => {
            // 副作用：计算单边的总宽度
            sideTotalWidth += button.width || defaultWidth

            // 设置未知宽度按钮的宽度
            if (!button.width) {
              return Object.assign(
                {},
                button,
                {
                  width: defaultWidth,
                }
              )
            }

            return button
          })

          return {
            [buttonsKey]: buttons,
            [widthKey]: sideTotalWidth,
          }
        }
        let initData = null

        // 如果未传入按钮数组，那么不需要以下初始化！
        if (!this.data[buttonsKey]) {
          return
        }

        if (!this.data[widthKey]) {
          // case1.1 如果传入 (left|right)ButtonWidth 会重新计算 (left|right)Width
          // case1.2 如果什么都没有设置，根据最小值初始化按钮列表数据
          initData = getInitData(this.data[buttonWidthKey] || MIN_BTN_W)
        } else {
          // case2. 如果传入 (left|right)Width 会平均剩余给未设置宽度的按钮
          let sideTotalWidth = 0
          const widthUnknownButtons = this.data[buttonsKey].filter(button => {
            // 副作用：计算宽度已知的按钮们占用的宽度
            if (button.width) {
              sideTotalWidth += button.width
            }

            return !button.width
          })

          // 如果已设置宽度按钮的总宽度超出可能，那么抛异常
          if (
            sideTotalWidth > this.data[widthKey] ||
            (sideTotalWidth === this.data[widthKey] &&
              widthUnknownButtons.length > 0)
          ) {
            throw new Error(
              `You should set enough "${widthKey}" for <wechat-swipeout> component`
            )
          }

          if (widthUnknownButtons.length > 0) {
            // 如果存在未设置宽度的按钮，那么设置宽度和更新按钮
            const restButtonWidth =
              (this.data[widthKey] - sideTotalWidth) /
              widthUnknownButtons.length
            widthUnknownButtons.forEach(button => {
              button.width = restButtonWidth
            })
            initData = {
              [buttonsKey]: widthUnknownButtons,
            }
          } else if (sideTotalWidth !== this.data[widthKey]) {
            // 如果不存在未设置宽度的按钮，且宽度不完全匹配，那么调整宽度到正合适
            initData = {
              [widthKey]: sideTotalWidth,
            }
          }
        }

        // console.log('attached, setData :', initData) // TODO: test to del
        this.setData(initData)
      }

      // 接口方式（非 slot 方式）
      _initialize('left')
      _initialize('right')
    },
    // 计算总宽度
    calculateTotal() {
      const leftW = this.data.leftWidth || 0
      const rightW = this.data.rightWidth || 0
      const totalWidth = leftW + this.data.width + rightW
      // TODO: test to del
      // console.log(`calculateTotal: ${totalWidth} = ${leftW} + ${this.data.width} + ${rightW}`)

      this.setData({
        totalWidth,
      })
    },
    // 准备阈值 (px)
    prepareThreshold() {
      const leftWidthInIntPX = this.data.leftWidth / 750 * WINDOW_W
      const rightWidthInIntPX = this.data.rightWidth / 750 * WINDOW_W
      this._leftThreshold = leftWidthInIntPX / 2
      this._rightThreshold = rightWidthInIntPX / 2
    },
    // 开启动画效果
    setupAnimation() {
      this.setData({
        useAnimation: true
      })
    },
    // 记录 touch 初始位置
    onTouchStart() { /* this._startX = e.changedTouches[0].pageX */ },
    // touch 结束后滑动到位
    onTouchEnd() {
      // this._endX = e.changedTouches[0].pageX
      const { /* _startX, _endX, */ _leftThreshold, _rightThreshold } = this

      if (this.data.leftVisibleWidth > _leftThreshold) {
        this._openLeft()
      } else if (this.data.rightVisibleWidth > _rightThreshold) {
        this._openRight()
      } else {
        this._close()
      }
    },
    // FIXME: enhance ??? 节流？防抖？
    // 当滑动的时候，修改 按钮的 可视宽度 和 吸附状态
    onChange(e) {
      // (px)
      const originalX = e.detail.x
      const offset = this.data.leftWidth ? this.data.leftWidth / 750 * WINDOW_W : 0
      const x = originalX + offset
      // console.log(`onChange, original x is ${originalX} ; x is ${x}`) // TODO: test to del

      const data = {}

      // 如果右侧按钮存在，那么计算可视宽度和吸附状态
      if (this.data.rightButtons) {
        if (x < 0) {
          const rightWidthInIntPX = this.data.rightWidth / 750 * WINDOW_W
          data.rightVisibleWidth = Math.min(-x, rightWidthInIntPX)

          // 判断贴边状态
          const isRightSticked = -x > rightWidthInIntPX
          if (this.data.stickToRightEdge && this.data.isRightSticked !== isRightSticked) {
            data.isRightSticked = isRightSticked
          }
        } else {
          data.rightVisibleWidth = 0
        }
      }

      // 如果左侧按钮存在，那么计算可视宽度和吸附状态
      if (this.data.leftButtons) {
        if (x > 0) {
          const leftWidthInIntPX = this.data.leftWidth / 750 * WINDOW_W
          data.leftVisibleWidth = Math.min(x, leftWidthInIntPX)

          // 判断贴边状态
          const isLeftSticked = x > leftWidthInIntPX
          if (this.data.stickToLeftEdge && this.data.isLeftSticked !== isLeftSticked) {
            data.isLeftSticked = isLeftSticked
          }
        } else {
          data.leftVisibleWidth = 0
        }
      }

      // 修改允许出界状态
      /**
       * ImoNote:
       * 设置 canBeOut=false 的 case：
       * - case1. 如果 originalX 很小，小到小于 -_leftThreshold 意味着主内容右边快要全部露出来
       * - case2. 如果 originalX 很大，大到大于 -_rightThreshold 意味着主内容左边快要全部露出来
       * 下式 !(case1 || case2) 即为 canBeOut=true 的情况
       */
      data.canBeOut = !((originalX < -this._leftThreshold && !this.data.rightWidth) ||
        (originalX > -this._rightThreshold && !this.data.leftWidth))

      // 删除 data 中没变化的属性，如果有变化才更新
      let hasChanged = false
      Object.keys(data).forEach(key => {
        if (data[key] !== this.data[key]) {
          hasChanged = true
        } else {
          delete data[key]
        }
      })
      if (hasChanged) {
        // console.log('onChange, hasChanged data :', data) // TODO: test to del
        this.setData(data)
      }
    },
    // 点击按钮时，触发 press 自定义事件
    onButtonPress(e) {
      const hash = e.currentTarget.dataset.hash

      // 自动关闭按钮
      if (this.data.autoClose) {
        this._close()
      }

      const eventDetail = { hash }
      const eventOption = {
        bubbles: true,
      }
      this.triggerEvent('press', eventDetail, eventOption)
    },
    // (点击|释放)按钮的时候，反转背景色
    toggleBackgroundColor(e) {
      const getToggledButtons = (key, theHash) => {
        let hasChanged = false

        const buttons = this.data[key].map(button => {
          const { hash, backgroundColor, underlayColor } = button
          // 对匹配的按钮做背景色的变化
          if (hash === theHash && underlayColor !== backgroundColor) {
            hasChanged = true
            // ImoNote: 这里使用 `{ ...button, // 覆盖选项 }` 的写法会编译错误
            return Object.assign(
              {},
              button,
              // 交换背景色
              {
                backgroundColor: underlayColor,
                underlayColor: backgroundColor,
              }
            )
          }

          return button
        })

        return hasChanged ? buttons : null
      }
      const checkSame = (buttons1, buttons2, hash) => {
        const btn1 = buttons1.find(button => button.hash === hash)
        const btn2 = buttons2.find(button => button.hash === hash)

        let isHashedButtonSame = true
        Object.keys(btn1).forEach(key => {
          if (btn1[key] !== btn2[key]) {
            isHashedButtonSame = false
          }
        })

        // 如果两个数组中匹配 hash 的按钮 backgroundColor,underlayColor 两个属性正好相反、其它属性均一致的话，返回 true
        return isHashedButtonSame
      }

      const key = e.currentTarget.dataset.key // 决定哪侧 (left|right)Buttons
      const theHash = e.currentTarget.dataset.hash // 决定哪个按钮
      const originalButtons = this.data[key]
      const toggledButtons = getToggledButtons(key, theHash)

      if (toggledButtons) {
        this.setData({
          [key]: toggledButtons
        })

        // 设置一个定时器来判断 .button-hover 何时消失，消失时复原按钮的背景色
        // ImoNote: 使用默认的 .button-hover 无法 select 到元素，只能自定义 hover-class 了（详见 index.wxml）
        const intervalId = setInterval(() => {
          const query = wx.createSelectorQuery().in(this)
          query.select(`.${prefix}button.${prefix}underlay`).fields({
            dataset: true
          })
          // ImoNote: 即使用 `select` 返回的结果也是一个数组
          query.exec(([res]) => {
            // case1. 用户松开了按钮
            // case2. 用户 hover 了别的按钮
            // case1 || case2 都需要复原当前按钮的背景色
            if (res === null || (res && res.dataset.hash !== theHash)) {
              const againToggledButtons = getToggledButtons(key, theHash)
              const isTheHashedButtonSame = checkSame(
                originalButtons,
                againToggledButtons,
                theHash
              )
              if (isTheHashedButtonSame) {
                this.setData({
                  [key]: againToggledButtons
                })
              }
              clearInterval(intervalId)
            }
          })
        }, 300) // ImoNote: 时间间隔过小的话也 query 不到元素
      }
    },
    /* “私有”方法 */
    _close() {
      const mainAllShownX = -this.data.leftWidth / 750 * WINDOW_W // (px)
      this.setData({
        offsetX: mainAllShownX
      })
    },
    _openLeft() {
      const leftAllShownX = 0
      this.setData({
        offsetX: leftAllShownX
      })
    },
    _openRight() {
      const rightAllShownX = (-this.data.leftWidth - this.data.rightWidth) / 750 * WINDOW_W
      this.setData({
        offsetX: rightAllShownX
      })
    },
    /* 供父组件调用的 */
    close() {
      if (!this.data.disabled) {
        this._close()
      }
    },
    openLeft() {
      if (!this.data.disabled) {
        this._openLeft()
      }
    },
    openRight() {
      if (!this.data.disabled) {
        this._openRight()
      }
    },
  },
})
